### PR TITLE
Improve arrow nav (horizontal only)

### PR DESCRIPTION
**Pull request by ellatrix**
Same as #2296, but without repositioning vertically.

Fixes various browser issues with arrow key navigation. #2124
Fixes link boundaries. #2104

### CODE REVIEW

**Comment by iseulde** [comment]
I wonder if it's a good idea to add some static methods to Editable for Editable specific logic, and getting the next editor etc.

**Comment by youknowriad** [comment]
Several notes here:

 - What if a block contains an input, an editable and a textarea for example, It looks like the code is ignoring the inputs and textareas right?
 - What bothers me is that sometimes we're using `onFocus` to navigate and in other cases we use DOM or TinyMCE API. Can't we be consistent and always use DOM Api (or TinyMCE since I see it as a helper over DOM)?
 - Also, I'm seeing some inconsistent caret placemenet. Say I have a separator between two text blocks. Depending on whether the first text block is empty/full the caret won't be on the same position if I navigate down to the text block after the tabbable.

Basically, What I'm proposing is something like this:

```js
if ( isEdge( currentTabbable ) ) {
   const nextTabbable = getNextTabbable( currentTabbable );
   focusBeginingOrEndOfNextTabbbable( nextTabbable );
}
```

with TinyMCE as a special case of "tabbable", and handled like a special case inside these functions. Is it possible this way?

**Comment by iseulde** [comment]
> What bothers me is that sometimes we're using onFocus to navigate and in other cases we use DOM or TinyMCE API. Can't we be consistent and always use DOM Api (or TinyMCE since I see it as a helper over DOM)?

That's what it's doing, using the TinyMCE API. But we have to fall back to onFocus if there is no editable (or nothing at all) in the next block, no?

> Also, I'm seeing some inconsistent caret placemenet. Say I have a separator between two text blocks. Depending on whether the first text block is empty/full the caret won't be on the same position if I navigate down to the text block after the tabbable.

Hm, I'm not sure how to reproduce.

> Basically, What I'm proposing is something like this:

Sounds good to me. Sounds like it's the same thing but slightly abstracted.

**Comment by iseulde** [comment]
Do you think we should do #2424 first?

**Comment by iseulde** [comment]
I do feel like we should try to handle keys at the source. So Editable or a modified textarea would signal when to handle focus. Instead of a wrapper around all of it which tries to handle the keys without context.

**Comment by youknowriad** [comment]
> That's what it's doing, using the TinyMCE API. But we have to fall back to onFocus if there is no editable (or nothing at all) in the next block, no?

Understood, I was thinking we could consider the block's wrapper as a "tabbable" but the question would be how we decide what's tabbable and what's not? I also raise this concern because the component in #2424 would be less "generic", it will be aware of blocks and state... Maybe we could resolve this by providing a prop `extraTabbableSelector` or something. 

> Hm, I'm not sure how to reproduce.

Me neither :) I don't know what happened

> Do you think we should do #2424 first?

I'm fine either way :) we could also just extract the behavior in this PR

**Comment by iseulde** [comment]
What if we use HOCs and force everything "tabbable" in the editor to be wrapped? FlowedTextarea, FlowedInput, FlowedEditableList etc. (I'm also bad with naming.) Then we know exactly what we have. If the block doesn't have anything, we use the wrapper of the block.

File: editor/modes/visual-editor/block.js [hunk 1]
```diff
@@ -251,12 +314,54 @@ class VisualEditorBlock extends Component {
...
+					return;
+				}
+
+				followingEditor.focus();
+
+				if ( reverse ) {
+					followingEditor.selection.select( followingEditor.getBody(), true );
+					followingEditor.selection.collapse( false );
+				} else {
+					followingEditor.selection.setCursorLocation();
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: What does this do? Can we comment as such?

File: editor/modes/visual-editor/block.js [hunk 2]
```diff
@@ -251,12 +314,54 @@ class VisualEditorBlock extends Component {
...
+						onFocus( followingBlock.uid, { offset: reverse ? -1 : 0 } );
+					}
+
+					return;
+				}
+
+				followingEditor.focus();
+
+				if ( reverse ) {
+					followingEditor.selection.select( followingEditor.getBody(), true );
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: Is the reason we get TinyMCE instance primarily to have access to its APIs for selecting the end of the next block? Native APIs aren't quite as nice, but not absurd:

```js
const range = document.createRange();
range.selectNodeContents( followingEditor );
range.collapse( false );
const selection = window.getSelection();
selection.removeAllRanges();
selection.addRange( range );
```

I think I might be more okay with this when we pull out and isolate these behaviors, as in #2424.

File: editor/modes/visual-editor/block.js [hunk 3]
```diff
@@ -251,12 +314,54 @@ class VisualEditorBlock extends Component {
 
 	onKeyDown( event ) {
 		const { keyCode, target } = event;
+		const { previousBlock, nextBlock, onFocus } = this.props;
+		const up = keyCode === UP;
+		const down = keyCode === DOWN;
+		const left = keyCode === LEFT;
+		const right = keyCode === RIGHT;
+
+		if ( up || down || left || right ) {
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: Pyramid of `if` could be improved if we delegated this to some helper methods that had early returns (as previously with `handleArrowKey`). See below comment too where we have early return that will inadvertently skip other `onKeyDown` logic.

File: editor/modes/visual-editor/block.js [hunk 4]
```diff
@@ -251,12 +314,54 @@ class VisualEditorBlock extends Component {
...
+					onFocus( followingBlock.uid, { offset: reverse ? -1 : 0 } );
+				}
+			}
+
+			if ( editor ) {
+				const reverse = up || left;
+				const followingBlock = reverse ? previousBlock : nextBlock;
+
+				if ( ! isEdge( { editor, reverse } ) ) {
+					return;
```
Comment (aduth) [inline editor/modes/visual-editor/block.js]: A `return` here makes me wary since we don't isolate this logic from other `onKeyDown` handlers. Currently the only remaining logic which would be skipped is the `ENTER` handling, that would never coincide with this, but seems easy to overlook in future maintenance.

**Review by aduth — COMMENTED** [review]
In Firefox, when moving to previous or next editable by arrow key, the focused block changes, but the editable is not focused, so I lose the text caret.

**Comment by iseulde** [comment]
An alternative PR was merged.
